cadc-inventory-db/src/main/java/org/opencadc/inventory/ArtifactSyncWorker.java

```diff
@@ -0,0 +1,143 @@
+*  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+*  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+*  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+*  General Public License for           Générale Publique GNU Affero
+*  more details.                        pour plus de détails.
+*
+*  You should have received             Vous devriez avoir reçu une
+*  a copy of the GNU Affero             copie de la Licence Générale
+*  General Public License along         Publique GNU Affero avec
+*  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+*  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+*                                       <http://www.gnu.org/licenses/>.
+*
+************************************************************************
+ */
+
+package org.opencadc.inventory;
+
+import ca.nrc.cadc.db.TransactionManager;
+import ca.nrc.cadc.io.ResourceIterator;
+import java.io.IOException;
+import org.apache.log4j.Logger;
+import org.opencadc.inventory.db.ArtifactDAO;
+import org.opencadc.inventory.db.HarvestState;
+import org.opencadc.inventory.db.HarvestStateDAO;
+import org.opencadc.vospace.DataNode;
+import org.opencadc.vospace.db.NodeDAO;
+
+/**
+ * This class performs the work of synchronizing the size of Data Nodes between backend storage and Node Persistence
+ * 
+ * @author adriand
+ */
+public class ArtifactSyncWorker implements Runnable {
+    private static final Logger log = Logger.getLogger(ArtifactSyncWorker.class);
+
+    private final HarvestState harvestState;
+    private final NodeDAO nodeDAO;
+    private final ArtifactDAO artifactDAO;
+    private final HarvestStateDAO harvestStateDAO;
+    private final Namespace storageNamespace;
+
+    public ArtifactSyncWorker(HarvestStateDAO harvestStateDAO, HarvestState harvestState, ArtifactDAO artifactDAO,
+                              Namespace namespace) {
+        this.harvestState = harvestState;
+        this.harvestStateDAO = harvestStateDAO;
+        this.nodeDAO = new NodeDAO(harvestStateDAO);
+        this.artifactDAO = artifactDAO;
+        this.storageNamespace = namespace;
+    }
+
+    @Override
+    public void run() {
+        log.debug("Start harvesting " + harvestState.toString() + " at " + harvestState.curLastModified);
+
+        TransactionManager tm = nodeDAO.getTransactionManager();
+        try (final ResourceIterator<Artifact> iter = artifactDAO.iterator(storageNamespace, null,
+                harvestState.curLastModified, true)) {
+            while (iter.hasNext()) {
+                Artifact artifact = iter.next();
+                DataNode node = nodeDAO.getDataNode(artifact.getURI());
+                if ((node != null) && !artifact.getContentLength().equals(node.bytesUsed)) {
+                    node.bytesUsed = artifact.getContentLength();
+                    tm.startTransaction();
+                    try {
+                        nodeDAO.put(node);
```

Comment:
Inside the txn but before the put(node), this should lock the the node before updating in case state changed since the getDataNode call.

The lock method returns another instance of the node that might be different from getDataNode if another process also updated it (eg NodePersistenceImpl.get(...) if we do that "cache update " or validation) and it might return null if the node was deleted... all part of race condition handling (but not easily testable without a lot of effort that's not worth it). Once you have started and txn and locked the node, there is not even much point in rechecking the bytesUsed: just assign the value and put. The metaChecksum will optimize whether the update is needed or not.

-----

for harvestStateDAO.put(...) you do not need the second argument to force the timestamp update. Here it should just rely on changing the fields of the object.

Logically, the harvestStateDAO.put(...) needs to happen for every artifact event handled, not just the ones that cause an update so it needs to go outside the `if (node size needs update)`. That's OK because the harvestStateDAO.put(...) call can happen outside the transaction (it risks re-evaluating an artifact event, which is idempotent so harmless). Because I will probably configure the HarvestStateDAO to buffer updates and possibly even run the maintenance it *must* be outside the transaction because that happens inside the put().